pkgs/stdenv/generic/default.nix

```diff
@@ -64,9 +73,57 @@ let
             { allow${unfreeOrBroken} = true; }
           to ~/.nixpkgs/config.nix.
         '';
+
+      # Check whether unfree packages are allowed and if not, whether the
+      # package has an unfree license and is not explicitely allowed by the
+      # `allowUNfreePredicate` function.
+      hasDeniedUnfreeLicense = attrs:
+        !allowUnfree &&
+        isUnfree (lib.lists.toList attrs.meta.license or []) &&
+        !allowUnfreePredicate attrs;
+
+      # Check whether two sets are mutual exclusive
+      mutualExclusive = a: b:
+        (builtins.length a) == 0 ||
+        (!(builtins.elem (builtins.head a) b) &&
+         mutualExclusive (builtins.tail a) b);
+
+      # Check whether an package has the license set
+      licenseCheckable = attr:
+        builtins.hasAttr "meta" attrs && builtins.hasAttr "license" attrs.meta;
+
+      # Check whether the license of the package is whitelisted.
+      # If the package has no license, print a warning about this and allow the
+      # package (return that it is actually whitelisted)
+      hasWhitelistedLicense = attrs:
+        if licenseCheckable attrs then
+          builtins.elem attrs.meta.license whitelistedLicenses
+        else
+          #builtins.trace "Has no license: ${attrs.name}, allowing installation"
+                         true;
+
+      # Check whether the license of the package is blacklisted.
+      # If the package has no license, print a warning about this and allow the
+      # package (return that it is actually not blacklisted)
+      hasBlacklistedLicense = attrs:
+        if licenseCheckable attrs then
+          builtins.elem attrs.meta.license blacklistedLicenses
+        else
+          #builtins.trace "Has no license: ${attrs.name}, allowing installation"
+                         false;
+
     in
-    if !allowUnfree && isUnfree (lib.lists.toList attrs.meta.license or []) && !allowUnfreePredicate attrs then
-      throwEvalHelp "Unfree" "has an unfree license"
+    if !(mutualExclusive whitelistedLicenses blacklistedLicenses) then
```

Comment:
Done in https://github.com/NixOS/nixpkgs/commit/0feb19b6b498c7d494055c0c97fc38c052f595a3